### ISSUE TITLE
Diet Tracker

### ISSUE THREAD

**Issue by ChrisPichlerEMU** [COSC481W-2024Fall/Ript-Fitness#5]
As a user of the Ript Fitness app, I want to be able to track my macronutrient intake on a daily basis based on foods that I eat throughout a specific day.

Feature name - Diet Tracker
Description - Tracks a user's daily macronutrient intake based on foods that the user inputs that will be stored into the database. The user can keep track of each day's intake by defining what foods in the food database were eaten in a specific day. The app will calculate the total macronutrient intake based on the foods defined in each day and will return it to the front end to be displayed to the user.

Task list - 
1. Define a controller with endpoints for CRUD operations for Food objects as well as Day objects (Add a Food object, edit a Food object, get a Food object's stats, delete a Food object, add a Day object, add Foods to a Day object, get a Day object's total stats, delete Foods in a Day object, delete a Day object).
2. Define service classes that use the controller endpoints to incorporate the business logic to actually do the logic for the endpoints to successfully return the correct data to the front end.
3. Define database queries in a Repository interface to manipulate the database based on the endpoint called and the data input with the HTTP request
4. Write unit/integration tests to meet the 70%+ code coverage required to merge into the main branch.
5. Confirm that code has been successfully deployed on Azure once code is merged into main.

Acceptance tests - 
1. If a user attempt to add a food to the database, they are able to click a button to save it, and the backend will take in that Food object and add each of the nutrient values to the database as they were typed by the user.
2. If a user wants to edit a food, they can click the correct button on the UI to edit the food, and the backend will take in that new Food object and successfully edit it so that the new values are stored in the database.
3. If a user deletes a food, that food will not be returned by any findBy queries sent to the database.
4. If a user wishes to add a day that holds a list of foods, they can click the corresponding button on the UI and a new row will be added to the Day database.
5. If a user wishes to add a food that was eaten in a specific day, that user can click the corresponding button on the UI and add any number of foods to that day, and the backend will return the correct total nutrient values based on the foods in that day.
6. If a user wishes to delete a food that was kept in a specific day, that user can click the correct button in the UI and delete any foods in that day's food list. The correct totals will then be returned to the front end based on those foods no longer being eaten in that day.

Acceptance criteria - 
1. A user can successfully add, edit, get, and delete a Food object from the database by targeting the correct endpoints.
2. A user can successfully add, edit, get, and delete a Day object from the database by targeting the correct endpoints.
3. The values returned in a Day object from the endpoints are the correct totals for each macronutrient calculated.

- User (Persona) - Bo, who wants to calculate his macronutrient intake on a specific day.
- Feature (app is going to implement) - The app needs to display the data returned from the HTTP endpoint
- Business value - The user doesn't need to calculate their daily micronutrient intake by hand or by using Excel.

HTTP Endpoints: 
1. Add a food:
     Type: POST
     URL: localhost:8080/addFood
     Body: FoodDto object

2. Get a food:
     Type: GET
     URL: localhost:8080/getFood/{insert Food name here}
     Body: None

3. Edit a food:
     Type: PUT
     URL: localhost:8080/editFood/{insert Food name here}
     Body: FoodDto object that will replace the current object in the database

4. Delete a food:
     Type: DELETE
     URL: localhost:8080/deleteFood/{insert Food name here} 
     Body: None

5. Add a day:
     Type: POST
     URL: localhost:8080/addDay
     Body: DayDto object to be added

6. Get a day (calculates total stats):
     Type: PUT
     URL: localhost:8080/getDay/{insert Day ID here}
     Body: None

7. Delete a day
     Type: DELETE
     URL: localhost:8080/deleteDay/{insert Day ID here}
     Body: None

8. Add Foods to a Day:
     Type: POST
     URL: localhost:8080/addFoodsToDay/{insert Day ID here}
     Body: Food IDs in array format ([1, 2, 3]) to be added to Day

9. Delete Foods in a Day:
     Type: POST
     URL: localhost:8080/deleteFoodsInDay/{insert Day ID here}
     Body: Food IDs in array format ([1, 2, 3]) to be deleted from Day

Estimated time to completion = < 1 week
